File: Python/Data Structures & Algorithms Module - Python/Week6/LinkedStack.py
"""
LinkedStack.py
Billy Cussen
16/11/2020
"""

class Node:
    def __init__(self, element, next):
        self.element = element
        self.next = next

class LinkedStack:
    def __init__(self):
        self.first = None

    def isEmpty(self):
        return self.first == None

    def getSize(self):
        return self.getSizeHelper(self.first)

    def getSizeHelper(self, curr):
        if curr == None:
            return 0
        return 1 + self.getSizeHelper(curr.next)

    def toString(self):
        if self.isEmpty():
            raise Exception("This Stack is Empty, can't traverse!")
        position = 0
        result = "***LINKED STACK TRAVERSAL***\n"
        return self.toStringHelper(self.first,result,position)

    def toStringHelper(self, curr, result, position):
        if curr == None:
            result+="***END***\n"
            return result
        result += "Position "+str(position)+", Element "+str(curr.element)+"\n"
        position+=1
        return self.toStringHelper(curr.next,result,position)

    def push(self,elem):
        if self.isEmpty():
            self.first = Node(elem, None)
        else:
            self.first = Node(elem, self.first)

    def pop(self):
        if self.isEmpty():
            raise Exception("This Stack is Empty, can't pop!")
        else:
            elem = self.first.element
            self.first = self.first.next
            return elem

    def peek(self):
        if self.isEmpty():
            raise Exception("This Stack is Empty, can't peek!")
        else:
            return self.first.element

myLinkedStack = LinkedStack()

try:
    print("Is Empty: "+str(myLinkedStack.isEmpty()))
except Exception as e:
    print(e)
finally:
    try:
        print("Size: "+str(myLinkedStack.getSize()))
    except Exception as e:
        print(e)
    finally:
        try:
            print("Peek: "+str(myLinkedStack.peek()))
        except Exception as e:
            print(e)
        finally:
            try:
                print("Pop: "+str(myLinkedStack.pop()))
            except Exception as e:
                print(e)
            finally:
                try:
                    print(myLinkedStack.toString())
                except Exception as e:
                    print(str(e)+"\n")

myLinkedStack.push(2)
myLinkedStack.push(4)
myLinkedStack.push(6)
myLinkedStack.push(8)
myLinkedStack.push(10)
myLinkedStack.push(12)

try:
    print("Is Empty: "+str(myLinkedStack.isEmpty()))
except Exception as e:
    print(e)
finally:
    try:
        print("Size: "+str(myLinkedStack.getSize()))
    except Exception as e:
        print(e)
    finally:
        try:
            print("Peek: "+str(myLinkedStack.peek()))
        except Exception as e:
            print(e)
        finally:
            try:
                print("Pop: "+str(myLinkedStack.pop()))
            except Exception as e:
                print(e)
            finally:
                try:
                    print(myLinkedStack.toString())
                except Exception as e:
                    print(e)

while myLinkedStack.isEmpty() == False:
    print("Pop: "+str(myLinkedStack.pop()))

try:
    print("Is Empty: "+str(myLinkedStack.isEmpty()))
except Exception as e:
    print(e)
finally:
    try:
        print("Size: "+str(myLinkedStack.getSize()))
    except Exception as e:
        print(e)
    finally:
        try:
            print("Peek: "+str(myLinkedStack.peek()))
        except Exception as e:
            print(e)
        finally:
            try:
                print("Pop: "+str(myLinkedStack.pop()))
            except Exception as e:
                print(e)
            finally:
                try:
                    print(myLinkedStack.toString())
                except Exception as e:
                    print(e)

File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week6/LinkedStack.java
/*
LinkedStack.java
Billy Cussen
16/11/2020
*/
public class LinkedStack<T>{

    Node<T> first;

    public static void main(String [] args){
        LinkedStack<Integer> myLinkedStack = new LinkedStack<Integer>();

        try{
            System.out.println("Is Empty: "+myLinkedStack.isEmpty());
        } catch(NullPointerException e){
            System.out.println(e.getMessage());
        } finally{
            try{
                System.out.println("Size: "+myLinkedStack.getSize());
            } catch(Exception e){
                System.out.println(e.getMessage());
            } finally{
                try{
                    System.out.println("Pop: "+myLinkedStack.pop());
                } catch(Exception e){
                    System.out.println(e.getMessage());
                } finally{
                    try{
                        System.out.println(myLinkedStack.toStringLinkedStack());
                    } catch(Exception e){
                        System.out.println(e.getMessage()+"\n");
                    } 
                }    
            }    
        }

        myLinkedStack.push(2);
        myLinkedStack.push(4);
        myLinkedStack.push(6);
        myLinkedStack.push(8);
        myLinkedStack.push(10);

        try{
            System.out.println("Is Empty: "+myLinkedStack.isEmpty());
        } catch(NullPointerException e){
            System.out.println(e.getMessage());
        } finally{
            try{
                System.out.println("Size: "+myLinkedStack.getSize());
            } catch(Exception e){
                System.out.println(e.getMessage());
            } finally{
                try{
                    System.out.println("Pop: "+myLinkedStack.pop());
                } catch(Exception e){
                    System.out.println(e.getMessage());
                } finally{
                    try{
                        System.out.println(myLinkedStack.toStringLinkedStack());
                    } catch(Exception e){
                        System.out.println(e.getMessage());
                    } 
                }    
            }    
        }
        
        while(!myLinkedStack.isEmpty()){
            System.out.println("Pop: "+myLinkedStack.pop());
        }

        try{
            System.out.println("Is Empty: "+myLinkedStack.isEmpty());
        } catch(NullPointerException e){
            System.out.println(e.getMessage());
        } finally{
            try{
                System.out.println("Size: "+myLinkedStack.getSize());
            } catch(Exception e){
                System.out.println(e.getMessage());
            } finally{
                try{
                    System.out.println("Pop: "+myLinkedStack.pop());
                } catch(Exception e){
                    System.out.println(e.getMessage());
                } finally{
                    try{
                        System.out.println(myLinkedStack.toStringLinkedStack());
                    } catch(Exception e){
                        System.out.println(e.getMessage());
                    } 
                }    
            }    
        }

    }

    void push(T element){
        if(isEmpty()){
            this.first = new Node<T>(element, null);
        } else{
            this.first = new Node<T>(element,this.first);
        }
    }

    T pop(){
        if(isEmpty()){
            throw new NullPointerException("Can't Pop, the Stack is Empty!");
        } else{
            T element = this.first.element;
            this.first = this.first.next;
            return element;
        }
    }

    T peek(){
        if(isEmpty()){
            throw new NullPointerException("Can't Peek, the Stack is Empty!");
        } else{
            return this.first.element;
        }
    }

    int getSize(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Size, the Stack is Empty!");
        } else{
            return sizeHelper(this.first);
        }
    }

    int sizeHelper(Node<T> current){
        if(current == null){
            return 0;
        }
        return 1 + sizeHelper(current.next);
    }

    boolean isEmpty(){
        return first == null;
    }

    String toStringLinkedStack(){
        if(isEmpty()){
            throw new NullPointerException("Can't Traverse, the Stack is Empty!");
        } else{
            String result = "***LINKED STACK TRAVERSAL***\n";
            int position = 0;
            return toStringLinkedStackHelper(this.first, result, position);
        }
    }

    String toStringLinkedStackHelper(Node<T> current, String result, int position){
        if(current == null){
            return result+"***END***\n";
        }
        result+="Position "+position+", Element: "+current.element+"\n";
        return toStringLinkedStackHelper(current.next, result, ++position);
    }
}

class Node<T>{

    T element;
    Node<T> next;

    Node(T element, Node<T> next){
        this.element = element;
        this.next = next;
    }

}